src/libcore/array/mod.rs

```diff
@@ -387,6 +387,84 @@ where
     }
 }
 
+/// Non-zero numbers
+///
+/// We require following from this trait:
+///  - [(); 0] does **not** implement NonZero.
+///  - [(); N] where N != 0 **implements** NonZero.
+///  As side effect of selected approach, this trait is also implemented
+///  for all other types, but it does not matter.
+#[unstable(feature = "array_default_internals", reason = "implementation detail", issue = "none")]
+pub auto trait NonZero {}
+#[unstable(feature = "array_default_internals", reason = "implementation detail", issue = "none")]
+impl !NonZero for [(); 0] {}
+
+// This is implementation for all non-zero-sized arrays
+#[stable(since = "1.4.0", feature = "array_default")]
+impl<T: Default, const N: usize> Default for [T; N]
+where
+    [(); N]: NonZero,
+{
```

Comment:
Presumably you want to add the typical `LengthAtMost32` bound, otherwise `[T; 33]` would instantly have a stable `Default` implementation.

Reply:
https://github.com/rust-lang/rust/pull/74060 It seems to me like `LengthAtMost32` will be gone soon, though I do think
it is a good idea to keep `LengthAtMost32` for one or two releases here in case this causes problems

Reply:
I added LengthAtMost32 bound back.